llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,359 @@
+#include "llvm/SYCLLowerIR/PropagateAspectUsage.h"
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/DiagnosticPrinter.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+
+using namespace llvm;
+
+namespace {
+
+class SYCLPropagateAspectUsageLegacyPass : public ModulePass {
+public:
+  static char ID; // Pass identification, replacement for typeid
+  SYCLPropagateAspectUsageLegacyPass() : ModulePass(ID) {
+    initializeSYCLPropagateAspectUsageLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  PropagateAspectUsagePass Impl;
+};
+
+} // namespace
+
+char SYCLPropagateAspectUsageLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLPropagateAspectUsageLegacyPass, "PropagateAspectUsage",
+                "Propagates aspect usage", false, false)
+
+ModulePass *llvm::createPropagateAspectUsagePass() {
+  return new SYCLPropagateAspectUsageLegacyPass();
+}
+
+namespace {
+
+using AspectsTy = SmallSet<int, 4>;
+using TypesWithAspectsTy = std::unordered_map<Type *, AspectsTy>;
+
+TypesWithAspectsTy GetTypesWithAspectsFromMetadata(Module &M) {
+  NamedMDNode *Node = M.getNamedMetadata("intel_types_that_use_aspects");
+  TypesWithAspectsTy Result;
+  if (!Node)
+    return Result;
+
+  LLVMContext &C = M.getContext();
+  for (size_t i = 0; i != Node->getNumOperands(); ++i) {
+    MDNode *N = Node->getOperand(i);
+    assert(N->getNumOperands() > 1 && "intel_types_that_use_aspect metadata "
+                                      "shouldn't contain empty metadata nodes");
+
+    MDString *TypeName = cast<MDString>(N->getOperand(0));
+    Type *T = StructType::getTypeByName(C, TypeName->getString());
+    if (!T) {
+      // TODO: warn?
+      continue;
+    }
+
+    AspectsTy &Aspects = Result[T];
+    for (size_t j = 1; j != N->getNumOperands(); ++j) {
+      Constant *C = cast<ConstantAsMetadata>(N->getOperand(j))->getValue();
+      Aspects.insert(cast<ConstantInt>(C)->getSExtValue());
+    }
+  }
+
+  return Result;
+}
+
+using TypesEdgesTy = std::unordered_map<Type *, std::vector<Type *>>;
+
+void PropagateAspectsThroughTypes(TypesEdgesTy &Edges, Type *Start,
+                                  AspectsTy &Aspects,
+                                  TypesWithAspectsTy &ResultAspects) {
+  std::unordered_set<Type *> Visited;
+  std::queue<Type *> queue;
+  queue.push(Start);
+  while (!queue.empty()) {
+    Type *T = queue.front();
+    queue.pop();
+    if (Visited.count(T))
```

Comment:
Consider the following example:
```
1->2
|  |
|  v
-->3
```
If we start from 1 then we push 2 and 3 into queue. Then we extract 2 and push 3 one more time since 3 is not present in Visited yet. Now we have two copies of vertex 3 in queue. For such cases the check in line 96 is introduced.

Reply:
Hm... Thank you for the explanation.